File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/3. Advanced Maths/javascript/03. Sieve of Eratosthenes.js
/**
QUESTION:
Given n, return how many primes are strictly less than n using the Sieve of Eratosthenes.

Approach:
Initialize an array of booleans set to true. Traverse from 2 up to n-1, and whenever a number is still marked prime, increment the count and mark all multiples starting from i*i as composite.

Time Complexity: O(n log log n)
Space Complexity: O(n)
*/

function countPrimes(n) {
    if (n <= 2) return 0;
    const primes = new Array(n).fill(true);
    let cnt = 0;
    for (let i = 2; i < n; i++) {
        if (primes[i]) {
            cnt++;
            for (let j = i * i; j < n; j += i) {
                primes[j] = false;
            }
        }
    }
    return cnt;
}

module.exports = { countPrimes };

File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/3. Advanced Maths/python/03. Sieve of Eratosthenes.py
"""QUESTION:
Given n, return the count of primes strictly less than n using the Sieve of Eratosthenes.

Approach:
Create a boolean list initialized to True. Traverse from 2 to n-1, and whenever a number remains marked prime, increment the count and mark its multiples starting from i*i as composite.

Time Complexity: O(n log log n)
Space Complexity: O(n)
"""

def count_primes(n: int) -> int:
    if n <= 2:
        return 0
    primes = [True] * n
    cnt = 0
    for i in range(2, n):
        if primes[i]:
            cnt += 1
            j = i * i
            while j < n:
                primes[j] = False
                j += i
    return cnt
